libkbfs/rekey_fsm.go

```diff
@@ -418,42 +434,41 @@ func NewRekeyFSM(fbo *folderBranchOps) RekeyFSM {
 }
 
 func (m *rekeyFSM) loop() {
-	for e := range m.reqs {
-		if e.eventType == rekeyShutdownEvent {
-			atomic.StoreInt32(&m.reqsClosed, 1)
-			close(m.reqs)
-		}
+	for {
+		select {
+		case e := <-m.reqs:
+			if e.eventType == rekeyShutdownEvent {
+				close(m.shutdownCh)
+			}
 
-		next := m.current.reactToEvent(e)
-		m.log.Debug("RekeyFSM transition: %T + %s -> %T",
-			m.current, e, next)
-		m.current = next
+			next := m.current.reactToEvent(e)
+			m.log.Debug("RekeyFSM transition: %T + %s -> %T",
+				m.current, e, next)
+			m.current = next
 
-		m.triggerCallbacksForTest(e)
+			m.triggerCallbacksForTest(e)
+			continue
+		default:
+		}
+		select {
+		case <-m.shutdownCh:
+			return
+		default:
+		}
 	}
 }
 
 // Event implements RekeyFSM interface for rekeyFSM.
 func (m *rekeyFSM) Event(event RekeyEvent) {
 	select {
-	case m.reqs <- event:
+	case <-m.shutdownCh:
 		return
 	default:
 	}
-	go func() {
-		// We use a spinning loop here to avoid doing a naked `m.reqs <- event`,
-		// which in racy conditions could cause a panic due to sending to closed
-		// channel. The spinning loop is fine here since reactToEvent
-		// implementations of states are supposed to consume events from m.reqs
-		// pretty fast.
-		for atomic.LoadInt32(&m.reqsClosed) == 0 {
-			select {
-			case m.reqs <- event:
-				return
-			default:
-			}
-		}
-	}()
+	select {
+	case <-m.shutdownCh:
+	case m.reqs <- event:
```

Comment:
> so in this select there's a probability that either an event is written or the shutdown is respected

Yes there is, but the event that's written will be consumed by the reader for sure (given that there are two selects in reader's loop). The `shutdownCh` case here is just to prevent it from blocking if it's already shutdown.

I don't feel like this approach either. We do need a buffered channel here as a queue. I guess an alternative is to have a mutex guarded `closed` variable; but that's no different than having the mutex guarding the channel (reference).